hack/main.go

```diff
@@ -105,97 +113,230 @@ func toSnakeCase(str string) string {
+					resourceType := resourceMethod.Type.Out(0)
+					r := resource{
+						Method: resourceMethod,
+						Type:   resourceType,
+					}
+					for _, operationMethod := range getMethods(resourceType) {
+						o := operation{
+							Method: operationMethod,
+						}
+						r.Operations = append(r.Operations, o)
+					}
+					c.Resources = append(c.Resources, r)
+				}
+			}
+			cs.Clients = append(cs.Clients, c)
+		}
+	}
+	return cs
+}
+
+func parseImports(cs clientset, packages ...string) []string {
+	imports := sets.NewString(packages...).Insert(cs.Type.PkgPath())
+	for _, c := range cs.Clients {
+		imports.Insert(c.Type.PkgPath())
+		for _, r := range c.Resources {
+			imports.Insert(r.Type.PkgPath())
+			for _, o := range r.Operations {
+				for _, i := range getIns(o.Method) {
+					if i.Kind() == reflect.Pointer {
+						i = i.Elem()
+					}
+					if i.PkgPath() != "" {
+						imports.Insert(i.PkgPath())
+					}
+				}
+				for _, i := range getOuts(o.Method) {
+					if i.Kind() == reflect.Pointer {
+						i = i.Elem()
+					}
+					if i.PkgPath() != "" {
+						imports.Insert(i.PkgPath())
+					}
+				}
+			}
+		}
+	}
+	return imports.List()
+}
+
+func executeTemplate(tpl string, cs clientset, folder string, packages ...string) {
+	tmpl := template.New("xxx")
+	tmpl.Funcs(
+		template.FuncMap{
+			"ToLower": func(in string) string {
+				return strings.ToLower(in)
+			},
+			"Quote": func(in string) string {
+				return `"` + in + `"`
+			},
+			"SnakeCase": func(in string) string {
+				return toSnakeCase(in)
+			},
+			"Args": func(in reflect.Method) []arg {
+				var out []arg
+				for i, a := range getIns(in) {
+					out = append(out, arg{
+						Type:       a,
+						IsVariadic: in.Type.IsVariadic() && i == in.Type.NumIn()-1,
+					})
+				}
+				return out
+			},
+			"Returns": func(in reflect.Method) []reflect.Type {
+				return getOuts(in)
+			},
+			"Pkg": func(in string) string {
+				return packageAlias(in)
+			},
+			"GoType": func(in reflect.Type) string {
+				return goType(in)
+			},
+		},
+	)
+	if tmpl, err := tmpl.Parse(tpl); err != nil {
+		panic(err)
+	} else {
+		if err := os.MkdirAll(folder, 0o755); err != nil {
+			panic(fmt.Sprintf("Failed to create directories for %s", folder))
+		}
+		file := "clientset.generated.go"
+		f, err := os.Create(path.Join(folder, file))
+		if err != nil {
+			panic(fmt.Sprintf("Failed to create file %s", path.Join(folder, file)))
+		}
+		if err := tmpl.Execute(f, map[string]interface{}{
+			"Clientset": cs,
+			"Packages":  parseImports(cs, packages...),
+		}); err != nil {
+			panic(err)
+		}
+	}
+}
+
+func generateMetricsWrapper(cs clientset, folder string, packages ...string) {
 	tpl := `
 package client
+{{- $clientsetPkg := Pkg .Clientset.Type.PkgPath }}
+{{- $metricsPkg := Pkg "github.com/kyverno/kyverno/pkg/metrics" }}
+{{- $discoveryPkg := Pkg "k8s.io/client-go/discovery" }}
+{{- $restPkg := Pkg "k8s.io/client-go/rest" }}
```

Comment:
Do we have restricted dependency on these folders? What do we need to keep in mind if we restructure the repo layout?

Reply:
What do you mean by _restricted dependency_ ?
I wanted to avoid magic names so I added a func to resolve the import alias from a package path.
If the package path changes we will need to update them here.

Reply:
Let's say I update the client code on my fork repo, will it look for `github.com/kyverno/kyverno/pkg/..` or `github.com/<forked-repo>/kyverno/pkg/..`?

Reply:
I think so, this is a local package as per your `go.mod`

Reply:
So it will look for the forked path?

Reply:
Yes, why not ?
The `go.mod` file in the forked repo says:
```
module github.com/kyverno/kyverno
```

Reply:
Thanks, wanted to confirm this.